src/benchmark_runner.cc

```diff
@@ -188,106 +243,91 @@ class BenchmarkRunner {
+    VLOG(2) << "Running " << b.name.str() << " for " << runForNumIters << "\n";
 
     std::unique_ptr<internal::ThreadManager> manager;
     manager.reset(new internal::ThreadManager(b.threads));
 
     // Run all but one thread in separate threads
     for (std::size_t ti = 0; ti < pool.size(); ++ti) {
-      pool[ti] = std::thread(&RunInThread, &b, iters, static_cast<int>(ti + 1),
-                             manager.get());
+      pool[ti] = std::thread(&RunInThread, &b, runForNumIters,
+                             static_cast<int>(ti + 1), manager.get());
     }
     // And run one thread here directly.
     // (If we were asked to run just one thread, we don't create new threads.)
     // Yes, we need to do this here *after* we start the separate threads.
-    RunInThread(&b, iters, 0, manager.get());
+    RunInThread(&b, runForNumIters, 0, manager.get());
 
     // The main thread has finished. Now let's wait for the other threads.
     manager->WaitForAllThreads();
     for (std::thread& thread : pool) thread.join();
 
-    IterationResults i;
     // Acquire the measurements/counters from the manager, UNDER THE LOCK!
     {
       MutexLock l(manager->GetBenchmarkMutex());
-      i.results = manager->results;
+
+      // Adjust real/manual time stats since they were reported per thread.
+      manager->results.real_time_used /= b.threads;
+      manager->results.manual_time_used /= b.threads;
+
+      c->NewResult(manager->results);
     }
 
     // And get rid of the manager.
     manager.reset();
 
-    // Adjust real/manual time stats since they were reported per thread.
-    i.results.real_time_used /= b.threads;
-    i.results.manual_time_used /= b.threads;
-    // If we were measuring whole-process CPU usage, adjust the CPU time too.
-    if (b.measure_process_cpu_time) i.results.cpu_time_used /= b.threads;
-
-    VLOG(2) << "Ran in " << i.results.cpu_time_used << "/"
-            << i.results.real_time_used << "\n";
-
-    // So for how long were we running?
-    i.iters = iters;
-    // Base decisions off of real time if requested by this benchmark.
-    i.seconds = i.results.cpu_time_used;
-    if (b.use_manual_time) {
-      i.seconds = i.results.manual_time_used;
-    } else if (b.use_real_time) {
-      i.seconds = i.results.real_time_used;
-    }
-
-    return i;
+    const ThreadManager::Result* res = std::prev(c->results_end());
+    VLOG(2) << "Ran in " << b.threads * res->cpu_time_used << "/"
+            << b.threads * res->real_time_used << "\n";
   }
 
-  IterationCount PredictNumItersNeeded(const IterationResults& i) const {
+ private:
+  IterationCount PredictNumItersNeeded() const {
     // See how much iterations should be increased by.
     // Note: Avoid division by zero with max(seconds, 1ns).
-    double multiplier = min_time * 1.4 / std::max(i.seconds, 1e-9);
+    double multiplier = min_time * 1.4 / std::max(c->Seconds(b), 1e-9);
```

Comment:
> DoNIterations could return the time taken and iteration count (instead of the full result) which could then be passed in here and used to determine the next count. That would be healthier, i think, than deferring that to the abstraction.

But that's *precisely* why it's not done that way. While it is true that it will produce
identical meaningful results for the current `IterationAveragingResultCollector` (since that
 is what currently being used); for the `NonAveragingBenchmarkRunner` `c->Seconds(b)`
would produce the total summary time, instead of some time of last run.

Am i missing the point here?

Reply:
i think i am, sorry. it's fine.